File: formcraft-plm/src/main/java/fr/formcraft/model/enums/ChangeRequestStatus.java
package fr.formcraft.model.enums;

/**
 * Engineering/Formulation Change Request (ECR/ECO) state machine.
 * Valid transitions: DRAFT -> SUBMITTED -> UNDER_REVIEW -> APPROVED -> IMPLEMENTED
 *                                                        \-> REJECTED -> DRAFT
 */
public enum ChangeRequestStatus {
    DRAFT,
    SUBMITTED,
    UNDER_REVIEW,
    APPROVED,
    REJECTED,
    IMPLEMENTED;

    public boolean canTransitionTo(ChangeRequestStatus target) {
        return switch (this) {
            case DRAFT -> target == SUBMITTED;
            case SUBMITTED -> target == UNDER_REVIEW;
            case UNDER_REVIEW -> target == APPROVED || target == REJECTED;
            case APPROVED -> target == IMPLEMENTED;
            case REJECTED -> target == DRAFT;
            case IMPLEMENTED -> false;
        };
    }
}

File: app/enums/change_request_status.py
import enum


class ChangeRequestStatus(str, enum.Enum):
    DRAFT = "DRAFT"
    SUBMITTED = "SUBMITTED"
    UNDER_REVIEW = "UNDER_REVIEW"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    IMPLEMENTED = "IMPLEMENTED"

    def can_transition_to(self, target: "ChangeRequestStatus") -> bool:
        match self:
            case ChangeRequestStatus.DRAFT:
                return target == ChangeRequestStatus.SUBMITTED
            case ChangeRequestStatus.SUBMITTED:
                return target == ChangeRequestStatus.UNDER_REVIEW
            case ChangeRequestStatus.UNDER_REVIEW:
                return target in (ChangeRequestStatus.APPROVED, ChangeRequestStatus.REJECTED)
            case ChangeRequestStatus.APPROVED:
                return target == ChangeRequestStatus.IMPLEMENTED
            case ChangeRequestStatus.REJECTED:
                return target == ChangeRequestStatus.DRAFT
            case ChangeRequestStatus.IMPLEMENTED:
                return False
